datasets/cbm_dataframe_parser.py

```diff
@@ -0,0 +1,386 @@
+        """
+        Windowed signals are constructed from 2D raw data.
+        Fast Fourier Transform performed on these signals.
+        """
+
+        new_sampling_rate = int(self.selected_sensor_sr / self.downsampling_ratio)
+
+        file_raw_data_sampled = scipy.signal.decimate(file_raw_data,
+                                                      self.downsampling_ratio, axis=0)
+
+        file_raw_data_windows = self.split_file_raw_data(
+            file_raw_data_sampled,
+            new_sampling_rate,
+            self.signal_duration_in_sec,
+            self.overlap_ratio
+            )
+
+        # First dimension: 3
+        # Second dimension: number of windows
+        # Third dimension: Window for self.duration_in_sec. 1000 samples for default settings
+        num_features = file_raw_data_windows.shape[0]
+        num_windows = file_raw_data_windows.shape[1]
+
+        fft_output_window_size = self.cnn_1dinput_len
+
+        file_cnn_signals = np.zeros((num_features, num_windows, fft_output_window_size))
+
+        # Perform FFT on each window () for each feature
+        for window in range(num_windows):
+            for feature in range(num_features):
+
+                signal_for_fft = file_raw_data_windows[feature, window, :]
+
+                fft_out = abs(fft(signal_for_fft))
+                fft_out = fft_out[:fft_output_window_size]
+
+                fft_out[:self.num_start_zeros] = 0
+                fft_out[-self.num_end_zeros:] = 0
+
+                file_cnn_signals[feature, window, :] = fft_out
+
+            file_cnn_signals[:, window, :] = file_cnn_signals[:, window, :] / \
+                np.sqrt(np.power(file_cnn_signals[:, window, :], 2).sum())
+
+        # Reshape from (num_features, num_windows, window_size) into:
+        # (num_windows, num_features, window_size)
+        file_cnn_signals = file_cnn_signals.transpose([1, 0, 2])
+
+        return file_cnn_signals
+
+    def create_common_empty_df(self):
+        """
+        Create empty dataframe
+        """
+        df = pd.DataFrame(columns=self.common_dataframe_columns)
+        return df
+
+    def __init__(self, root, d_type,
+                 transform=None,
+                 target_sampling_rate_Hz=2000,
+                 signal_duration_in_sec=0.25,
+                 overlap_ratio=0.75,
+                 eval_mode=False,
+                 label_as_signal=True,
+                 random_or_speed_split=True,
+                 num_end_zeros=10,
+                 num_start_zeros=3,
+                 train_ratio=0.8,
+                 main_df=None):
+
+        if d_type not in ('test', 'train'):
+            raise ValueError(
+                "d_type can only be set to 'test' or 'train'"
+                )
+
+        self.main_df = main_df
+        self.df_normals = self.main_df[main_df['label'] == 0]
+        self.df_anormals = self.main_df[main_df['label'] == 1]
+
+        self.normal_speeds_Hz = list(set(self.df_normals['speed']))
```

Comment:
set conversion should be done from the beginning (normal_speeds_Hz should be set instead of a list)
As you would use stepping-over speeds with a constant step size, equivalent speeds will disturb the speed split property.

Reply:
normal_speeds_Hz  is set as a 'set' and does not include duplicates.